components/brave_wallet/browser/nft_metadata_fetcher.cc

```diff
@@ -0,0 +1,376 @@
+    return;
+  }
+
+  absl::optional<GURL> url = DecodeMetadataUri(*metadata);
+  if (!url || !url.value().is_valid()) {
+    std::move(callback).Run("", mojom::SolanaProviderError::kParsingError,
+                            l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+    return;
+  }
+
+  FetchMetadata(*url, base::BindOnce(
+                          &NftMetadataFetcher::CompleteGetSolTokenMetadata,
+                          weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void NftMetadataFetcher::CompleteGetSolTokenMetadata(
+    GetSolTokenMetadataCallback callback,
+    const std::string& response,
+    int error,
+    const std::string& error_message) {
+  std::move(callback).Run(response, mojom::SolanaProviderError(error),
+                          error_message);
+}
+
+// static
+absl::optional<uint32_t> NftMetadataFetcher::DecodeUint32(
+    const std::vector<uint8_t>& input,
+    size_t& offset) {
+  if (offset >= input.size() || input.size() - offset < sizeof(uint32_t)) {
+    return absl::nullopt;
+  }
+
+  // Read bytes in little endian order.
+  base::span<const uint8_t> s =
+      base::make_span(input.begin() + offset, sizeof(uint32_t));
+  uint32_t uint32_le = *reinterpret_cast<const uint32_t*>(s.data());
+
+  offset += sizeof(uint32_t);
+
+#if defined(ARCH_CPU_LITTLE_ENDIAN)
+  return uint32_le;
+#else
+  return base::ByteSwap(uint32_le);
+#endif
+}
+
+// static
+// Expects a the bytes of a Borsh encoded Metadata struct (see
+// https://docs.rs/spl-token-metadata/latest/spl_token_metadata/state/struct.Metadata.html)
+// and returns the URI string in of the nested Data struct (see
+// https://docs.rs/spl-token-metadata/latest/spl_token_metadata/state/struct.Data.html)
+// as a GURL.
+absl::optional<GURL> NftMetadataFetcher::DecodeMetadataUri(
+    const std::vector<uint8_t> data) {
+  size_t offset = 0;
+  offset = offset + /* Skip first byte for metadata.key */ 1 +
+           /* Skip next 32 bytes for `metadata.update_authority` */ 32 +
+           /* Skip next 32 bytes for `metadata.mint` */ 32;
+
+  // Skip next field, metdata.data.name, a string
+  // whose length is represented by a leading 32 bit integer
+  auto length = DecodeUint32(data, offset);
+  if (!length) {
+    return absl::nullopt;
+  }
+  offset += static_cast<size_t>(*length);
+
+  // Skip next field, `metdata.data.symbol`, a string
+  // whose length is represented by a leading 32 bit integer
+  length = DecodeUint32(data, offset);
+  if (!length) {
+    return absl::nullopt;
+  }
+  offset += static_cast<size_t>(*length);
+
+  // Parse next field, metadata.data.uri, a string
+  length = DecodeUint32(data, offset);
+  if (!length) {
+    return absl::nullopt;
+  }
+
+  // Prevent out of bounds access in case length value incorrent
+  if (data.size() <= offset + *length) {
+    return absl::nullopt;
+  }
+  std::string uri =
+      std::string(data.begin() + offset, data.begin() + offset + *length);
```

Comment:
From the looks of the use of DecodeUint32 the make_span call inside there shouldn't create any memory corruption issues since the lifetime of the data outlives the lifespan of the length var here. This shouldn't be a concern from what I can tell.